### PR TITLE
✨ Add support for generating based on scraping the Github API.

### PR DESCRIPTION
**Pull request by coderanger**
Includes general support for repo hosts and a specific implementation for Github.

I'm not sold on the naming of "repo host" but I couldn't think of anything better to cover future expansion to GitLab, Bitbucket, etc.

### CODE REVIEW

**Comment by coderanger** [comment]
An example YAML:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: testing
  namespace: default
spec:
  generators:
  - repoHost:
      github:
        organization: myorg
        tokenRef:
          name: github-token
          key: token
      filters:
      - repositoryMatch: ^service-.*
        pathExists: kubernetes/kustomization.yaml
  template:
    metadata:
      name: '{{ repository }}'
    spec:
      source:
        repoURL: '{{ url }}'
        targetRevision: '{{ branch }}'
        path: kubernetes/
      project: default
      destination:
        server: https://kubernetes.default.svc
        namespace: default
```

File: pkg/generators/repo_host.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	return DefaultRepoHostRequeueAfterSeconds
+}
+
+func (g *RepoHostGenerator) GetTemplate(appSetGenerator *argoprojiov1alpha1.ApplicationSetGenerator) *argoprojiov1alpha1.ApplicationSetTemplate {
+	return &appSetGenerator.RepoHost.Template
+}
+
+func (g *RepoHostGenerator) GenerateParams(appSetGenerator *argoprojiov1alpha1.ApplicationSetGenerator, applicationSetInfo *argoprojiov1alpha1.ApplicationSet) ([]map[string]string, error) {
```
Comment (OmerKahani) [inline pkg/generators/repo_host.go:46]: I think it would b better to pass only the metadata, so the function get's only the relevant part and the whole spec

File: pkg/generators/repo_host_test.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	argoprojiov1alpha1 "github.com/argoproj-labs/applicationset/api/v1alpha1"
+	"github.com/argoproj-labs/applicationset/pkg/services/repo_host"
+)
+
+func TestRepoHostGetSecretRef(t *testing.T) {
```
Comment (OmerKahani) [inline pkg/generators/repo_host_test.go:16]: It would be nice to reafctor this to [table tests](https://github.com/golang/go/wiki/TableDrivenTests)
and to add a name for each case

**Review by OmerKahani — COMMENTED** [review]
Can you please also add docs?

Very nice generator and use case!

File: pkg/generators/repo_host.go (hunk 1 above)
Comment (coderanger) [reply pkg/generators/repo_host.go:46]: Passing the whole object does leave a lot of nice doors open though, especially having generators that set Status conditions or having more global config in the future.

File: pkg/generators/repo_host_test.go (hunk 2 above)
Comment (coderanger) [reply pkg/generators/repo_host_test.go:16]: Updated :)

**Comment by OmerKahani** [comment]
@coderanger try to rebase / merge master, #207 should have fixed the problem in "Verify generated manifests"

**Comment by coderanger** [comment]
A question raised from a friend: should this support discovering branches as well as repos. Right now its hardwired to only consider the default branch of a repo. Seems like a thing that could be added in the future, another filter and a flag for allBranches: true or something.

**Review by jgwest — CHANGES_REQUESTED** [review]
Really nice PR @coderanger , thanks for contributing this! I've got some suggestions around name changes, and some doc/comment tweaks below, but nothing major... let me know what you think :smile: .

File: api/v1alpha1/applicationset_types.go [hunk 3]
```diff
@@ -21,6 +21,12 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// Utility struct for a reference to a secret key.
+type SecretRef struct {
+	Name string `json:"name"`
```
[the remark this answers is not in the document]
Comment (coderanger) [reply api/v1alpha1/applicationset_types.go:26]: `name` is how this is handled in corev1 so I would probably leave it there? https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#secretkeyselector-v1-core

It looks like:

```yaml
secretRef:
  name: foo
  key: bar
```

File: pkg/generators/repo_host.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("error listing repos: %v", err)
+	}
+	params := make([]map[string]string, 0, len(repos))
+	for _, repo := range repos {
+		params = append(params, map[string]string{
+			"organization": repo.Organization,
+			"repository":   repo.Repository,
+			"url":          repo.URL,
+			"branch":       repo.Branch,
```
[the remark this answers is not in the document]
Comment (coderanger) [reply pkg/generators/repo_host.go:86]: In the struct and template I think `branch` is correct. It's kind of like it's generating a whole bunch of possible structs that then get filtered down. Right now we only generate one for each org/repo pair where branch == default branch. But in the future it would (if enabled since it's an extra API request) spit out one struct for each branch that exists and then you would use additional filters to narrow that down like before (or maybe you wouldn't if you want to deploy every branch).

File: api/v1alpha1/applicationset_types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (coderanger) [reply api/v1alpha1/applicationset_types.go:26]: Though as I'm thinking of other places that do this, secret volumes use `secretName` so no actual consistency either way. Whatever you prefer.

File: api/v1alpha1/applicationset_types.go [hunk 5]
```diff
@@ -63,9 +69,10 @@ type ApplicationSetTemplateMeta struct {
 
 // ApplicationSetGenerator include list item info
 type ApplicationSetGenerator struct {
-	List     *ListGenerator    `json:"list,omitempty"`
-	Clusters *ClusterGenerator `json:"clusters,omitempty"`
-	Git      *GitGenerator     `json:"git,omitempty"`
+	List     *ListGenerator     `json:"list,omitempty"`
+	Clusters *ClusterGenerator  `json:"clusters,omitempty"`
+	Git      *GitGenerator      `json:"git,omitempty"`
+	RepoHost *RepoHostGenerator `json:"repoHost,omitempty"`
```
[the remark this answers is not in the document]
Comment (coderanger) [reply api/v1alpha1/applicationset_types.go:75]: No strong feelings either way, I feel like none of them are clear enough to not need to go read the docs, and all of them are clear enough once you know what it's talking about. >_< It's weird that there's no common term for these, I had never really thought about it before.

File: api/v1alpha1/applicationset_types.go [hunk 6]
```diff
@@ -112,6 +119,38 @@ type GitFileGeneratorItem struct {
...
+type RepoHostGeneratorGithub struct {
+	// Github org to scan. Required.
+	Organization string `json:"organization"`
+	// The Github API URL to talk to. If blank, use https://api.github.com/.
+	API string `json:"api,omitempty"`
+	// Authentication token reference.
+	TokenRef *SecretRef `json:"tokenRef,omitempty"`
+}
+
+// RepoHostGeneratorFilter is a single repoisitory filter.
```
[the remark this answers is not in the document]
Comment (coderanger) [reply api/v1alpha1/applicationset_types.go:144]: I think re: the first that `filters: []` and `filters: [{}]` will both technically work as no-ops since the filter code is fairly modular. The `[{}]` would be weird though and I would be fine if that was a validation error at some point.

File: pkg/services/repo_host/github.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+	for {
+		githubRepos, resp, err := g.client.Repositories.ListByOrg(ctx, g.organization, opt)
+		if err != nil {
+			return nil, err
+		}
+		for _, githubRepo := range githubRepos {
+			repos = append(repos, &HostedRepo{
+				Organization: githubRepo.Owner.GetName(),
+				Repository:   githubRepo.GetName(),
+				URL:          githubRepo.GetSSHURL(), // TODO Config flag for CloneURL (i.e. https://)?
```
[the remark this answers is not in the document]
Comment (coderanger) [reply pkg/services/repo_host/github.go:52]: This always uses the SSH URL, some users might prefer the HTTPS URL for cloning (eg. using token authentication instead of SSH keys). I think this is the more common use case with Argo so it's probably the right default but Argo supports a lot of auth modes (basic auth, token, github app, TLS auth which I don't think GHE supports anyway). It should be configurable in the `github: {}` struct with maybe an overall default in the generator since SSH vs HTTPS is common for many providers.

File: docs/Generators.md [hunk 8]
```diff
@@ -370,3 +370,97 @@ Any `config.json` files found under the `cluster-config` directory will be param
...
+* `organization`: Required name of the Github organization to scan. If you have multiple orgs, use multiple generators.
+* `api`: If using Github Enterprise, the URL to access it.
+* `tokenRef`: A Secret name and key containing the Github access token to use for requests. If not specified, will make anonymous requests which have a lower rate limit.
+
+For label filtering, the repository topics are used.
+
+### Filters
+
+Filters allow selecting which repositories to generate for. Filters are additive, specifying none will template every repository and each filter added will pare that down.
+
```
[the remark this answers is not in the document]
Comment (coderanger) [reply docs/Generators.md:412]: This wouldn't be correct, it has to match every rule rather than at least one.

File: docs/Generators.md [hunk 9]
```diff
@@ -370,3 +370,97 @@ Any `config.json` files found under the `cluster-config` directory will be param
...
+        pathExists: kubernetes/kustomization.yaml
+        labelMatch: deploy-ok
+  template:
+  # ...
+```
+
+* `repositoryMatch`: A regexp matched against the repository name.
+* `pathExists`: A path within the repository that must exist. Can be a file or directory, but do not include the trailing `/` for directories.
+* `labelMatch`: A regexp matched against repository labels. If any label matches, the repository is included.
+
```
[the remark this answers is not in the document]
Comment (coderanger) [reply docs/Generators.md:436]: Added more words up above in the section text.

File: docs/Generators.md [hunk 10]
```diff
@@ -370,3 +370,97 @@ Any `config.json` files found under the `cluster-config` directory will be param
     Only JSON file parsing is currently supported. The work to add support for YAML files is [tracked here](https://github.com/argoproj-labs/applicationset/issues/106).
 
 As with other generators, clusters *must* already be defined within Argo CD, in order to generate Applications for them.
+
+## Repository Host Generator
```
[the remark this answers is not in the document]
Comment (coderanger) [reply docs/Generators.md:374]: Updated

File: docs/Generators.md [hunk 11]
```diff
@@ -370,3 +370,97 @@ Any `config.json` files found under the `cluster-config` directory will be param
...
+
+The RepoHost generator uses the API of an SCMaaS provider to discover repositories. This fits well with many repos following the same GitOps layout patterns such as microservices.
+
+### Github
+
+The Github mode uses the Github API to scan and organization in either github.com or Github Enterprise.
```
[the remark this answers is not in the document]
Comment (coderanger) [reply docs/Generators.md]: Added above since it's more re: the generator than the github piece.

**Comment by coderanger** [comment]
Updated those three todos so they now all exist and work :)

File: pkg/generators/repo_host.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jgwest) [reply pkg/generators/repo_host.go:86]: Good point, agreed, thanks!

File: docs/Generators.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (jgwest) [reply docs/Generators.md:436]: Gotcha, my original second sentence was incorrect, but that got my thinking about how I would expect users to approach this design.

TLDR: I am proposing changing the outer filters to be processed as a logical OR list (in order to be included a repository must match at least one filter), and maintaining the conditions within a filter as an AND list (eg all must match)

Mainly I am making this proposal due to how the current behaviour handles scenario 3, below.

#### Scenario one: the user specifies no filters

```yaml
generators:
  repoHost:
    github:
      organization: hi
      # (...)
```

My guess at user expected behaviour: none of the repositories in the organization are filtered out.
Actual behaviour: Matches the user's expected behaviour.



#### Scenario two: the user specifies one filter

```yaml
generators:
  repoHost:
    github:
      organization: hi
      filters:
        - repositoryMatch: my-repos*
```

User expected behaviour: only repositories that match that one filter are included
Actual behaviour: Same.



#### Scenario three: the user specifies multiple filters

```yaml
generators:
  repoHost:
    github:
      organization: hi
      filters:
        - repositoryMatch: my-repo1
        - repositoryMatch: my-repo2
```

Expected: Both repository `my-repo1` and `my-repo` will be included.
Actual: no repositories will be included (unless, if I am not mistaken, the user happens to have a repository named `my-repo1-my-repo2` :smile:, since the PR looks for the regexp to match a substring)


Now, a response to this might be to say that the user is doing it wrong, and that they should be using reg-exps for the above case. But my expectation is that this behaviour will trip up many more users than it will 'click' with, and that familiarity with regular expression construction is a less common skill than one might think.

#### Proposal

So my proposal would be to switch the outer filters to be processed via logic OR, while maintaining logical AND for the inner conditions.  This would mean that scenario 3 above would match (my guess at) the user's expected behaviour.

What do you think? Is there some other user scenario that becomes much more difficult to express with this approach? It's definitely a balancing act between guessing at user expectations, trying to maximize the expressive power of the filter, and keeping the schema as simple/straightforward as possible/necessary.

File: api/v1alpha1/applicationset_types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jgwest) [reply api/v1alpha1/applicationset_types.go:26]: Agreed, definitely no consistency, I went looking to see if there was any pre-existing idiom here that we could draw upon and likewise came up short. 

Let's go with `secretName` here, I think this makes it more obvious where that value is coming from (which is otherwise not indicated/implied via context).

File: docs/Generators.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (coderanger) [reply docs/Generators.md:436]: So the only reason I made it an array in the first place was specifically the `PathExists` test which takes just one path, so you would need multiple blocks to check more than one file and in that context ANDing made more sense. I agree it's confusing in aggregate though. I can change that to `PathsExist` which takes a `[]string` and then make the overall multi-filter be OR, or just remove multi-filter?

File: api/v1alpha1/applicationset_types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jgwest) [reply api/v1alpha1/applicationset_types.go:75]: Agreed, I raised the question on Argo slack https://cloud-native.slack.com/archives/C01U45M2SVB/p1620063133114900 (to join: https://argoproj.github.io/community/join-slack/), we'll see if folks have opinions either way :+1: .

File: docs/Generators.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (coderanger) [reply docs/Generators.md:436]: Also another thing to factor in, you can get an OR behavior now with multiple generators but that will make redundant API calls which is probably not great because rate limits are a thing.

File: api/v1alpha1/applicationset_types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (coderanger) [reply api/v1alpha1/applicationset_types.go]: Done

File: docs/Generators.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (OmerKahani) [reply docs/Generators.md:436]: I think that multi-filter with OR is good behavior for now. We don't want to make multiple calls with multiple generators

[the remark this answers is not in the document]
Comment (jgwest) [reply docs/Generators.md:436]: I would say yes to multi-filter with OR... as for PathExists, I'm fine with a string array (the user MAY be able to use a  regex to accomplish the same thing with a non-string-array PathExists, but that is probably just complicating things for them) 

Re: OR behaviour with multiple generators, that's true. My expectation is it will be pretty common for folks to want to target multiple repository patterns in the list of repositories to target, so we may as well keep it in at the generator level to make things easier for them, as well as reducing repitition. It's true that it doesn't necessarily increase the expressive power of the generator, but it does improve the ergonomics (and does make it easier to avoid hitting API rate limits, as you said).

And thanks for walking us through your thinking, and helping us better understand the problem space :+1:, I tend to find schema design is naturally a lot of cognitive trial and error, as we try ideas on for size and think through how this affects various user scenarios. I'm really looking forward to getting this in (let me know if I can help!) :smile: .

File: api/v1alpha1/applicationset_types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (coderanger) [reply api/v1alpha1/applicationset_types.go:75]: ScmProvider seems to have good vibes from the channel so :+1: for that. Will do a mass rename.

File: docs/Generators.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (coderanger) [reply docs/Generators.md:436]: Regex on path checks unfortunately isn't easy with the way the github API works so I made it only take single paths to check (it's probing them directly). Sounds good on the rest though, will reorg things a bit.

File: api/v1alpha1/applicationset_types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jgwest) [reply api/v1alpha1/applicationset_types.go:75]: Sounds good, and thanks for jumping and answering questions :smile: .

**Comment by coderanger** [comment]
Okay, I renamed everything to "SCM provider" as the overall subsystem name and reworked the multi-filter behavior to be simpler (the OR/AND approach discussed above).

**Comment by jgwest** [comment]
@coderanger - FYI if you see an E2E test failure in the PR check, it's *not* anything wrong with your code, it's this: https://github.com/argoproj-labs/applicationset/issues/227 (I'm working on a fix).

File: examples/scm-provider-generator/scm-provider-example.yaml [hunk 12]
```diff
@@ -0,0 +1,23 @@
+apiVersion: argoproj.io/v1alpha1
+kind: ApplicationSet
+metadata:
+  name: guestbook
+spec:
+  generators:
+  - scmProvider:
```
Comment (jgwest) [inline examples/scm-provider-generator/scm-provider-example.yaml:7]: When I apply the example under `examples/scm-provider-generator/scm-provider-example.yaml`, it works :tada: , but (at least on my version of) Argo CD it returns an error within the Application status:

`kubectl apply -f scm-provider-example.yaml`

```yaml
# [jgw@localhost docs]$ k get application/argocd-example-apps-guestbook -o yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  creationTimestamp: "2021-05-09T23:56:20Z"
  name: argocd-example-apps-guestbook
  namespace: argocd-e2e
  # (...)
spec:
  destination:
    namespace: guestbook
    server: https://kubernetes.default.svc
  project: default
  source:
    path: guestbook
    repoURL: git@github.com:argoproj/argocd-example-apps.git
    targetRevision: master
status:
  conditions:
  - lastTransitionTime: "2021-05-09T23:56:20Z"
    message: 'rpc error: code = Unknown desc = error creating SSH agent: "SSH agent requested but SSH_AUTH_SOCK not-specified"'
    type: ComparisonError
  health:
    status: Healthy
  reconciledAt: "2021-05-09T23:56:20Z"
  # (...)
```

This message, `SSH agent requested but SSH_AUTH_SOCK not-specified`, appears to be Argo CD's (somewhat misleading) way of saying that it doesn't have `git@github.com:argoproj/argocd-example-apps.git` added as a private repository (with SSH private key). Once I add the repository it works as expected.

However, since this `scm-provider-example.yaml` is a simple example, we probably want it to work "out of the box" for users (eg without them needing to go through SSH repo add steps), so I recommend adding a cloneProtocol of https to the example, which should alleviate this issue:
```
  - scmProvider:
      cloneProtocol: https
```
Any downside of doing so that you can thing of? Have you seen similar behaviour?

**Review by jgwest — CHANGES_REQUESTED** [review]
Looking great, one tweak to the example and good to go (unless you had anything else you wanted to add)

File: examples/scm-provider-generator/scm-provider-example.yaml (hunk 12 above)
Comment (coderanger) [reply examples/scm-provider-generator/scm-provider-example.yaml:7]: D'oh, good catch. Downside of having a global key set up in my dev environment :)

**Review by jgwest — APPROVED** [review]
Really nice, thanks for working on this! It adds a powerful new tool into the ApplicationSet tool belt, and should really enable folks doing GitOps over multiple repositories in an organization to best automate that using ApplicationSets :tada:.
